apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+but this holds true only under certain constraints. These constraints include requirements
+like refraining from accessing the tensor's data pointer and avoiding the use of
+`out=` variants of the operators.
+For the full list of constraints, see [this documentation](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md#composite-compliance).
+When all these hold for an operator, we say that it is composite compliant.
+
+Unfortunately, operators that claim to be composite may occasionally deviate
+from these constraints. While such deviations may not pose issues when utilizing
+plain eager PyTorch, they can lead to complications when using `torch.func`
+transformations.
+
+### Testing for Composite Compliance
+
+We have tests to ensure that operators tagged as composite are indeed composite
+compliant. We test this by creating a new subclass `CompositeCompliantTensor`
+that utilizes the `__torch_dispatch__` mechanism. This mechanism is invoked for
+all operators in the testing, enabling us to detect any non-compliant behavior
+exhibited by an operator.
+
+Our testing approach involves running tests on the actual [operator](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1446),
+as well as their [backward formula](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1459)
+and [forward AD formula](https://github.com/pytorch/pytorch/blob/40b2c796dcae5768ff5de279b13914d5948fd414/test/test_ops.py#L1476).
+Testing both the backward and forward formulas is crucial because we may encounter
+scenarios involving `vmap(vjp(fn))` or `vmap(jvp(fn))`.
+
+## Support for `chunk_size` in `vmap` and `jacrev`
+The computation of the Jacobian can be memory-intensive, and users have raised
+concerns about related issues, (e.g., [this one](https://github.com/pytorch/functorch/issues/680)).
+In response to these concerns, we have introduced a feature that allows for the
+calculation of `jacrev` and `vmap` in smaller, user-defined chunks, determined
+by the `chunk_size` argument. This adjustment serves to reduce the peak memory
+usage during the computation process. With this argument, users can specify the
+number of rows of the Jacobian to be computed at once. This enhancement was
+incorporated into the [jacrev PR](https://github.com/pytorch/pytorch/pull/89376)
+and the [vmap PR](https://github.com/pytorch/pytorch/pull/91157).
+
+## Support for `linearize` transform
+
+The "jvp" transform is designed to calculate both `f(x)` and the Jacobian-vector
+product. Consequently, even when one intends to compute the Jacobian-vector
+product for fixed inputs, the `jvp` transform still redundantly evaluates `f(x)`.
+To address such scenarios, the `linearize` transform comes into play. This
+transform proves valuable when multiple `jvp` computations are needed for
+constant `primals`.
```

Comment:
Should we define primals here?

Reply:
Would replacing it with `inputs` be ok? We do talk about it 


> Consequently, even when one intends to compute the Jacobian-vector product for fixed inputs, the `jvp` transform still redundantly evaluates `f(x)`.

Reply:
Yeah inputs is fine